rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,220 @@
+                    // If the end of the range is not byte-aligned, we need to read one more byte
+                    end_byte_offset += 1;
+                }
+
+                let start = self.buffer_offset + start_byte_offset;
+                let end = self.buffer_offset + end_byte_offset;
+                min = min.min(start);
+                max = max.max(end);
+
+                start..end
+            })
+            .collect::<Vec<_>>();
+
+        trace!(
+            "Scheduling I/O for {} ranges spread across byte range {}..{}",
+            byte_ranges.len(),
+            min,
+            max
+        );
+
+        let bytes = scheduler.submit_request(byte_ranges);
+
+        let bits_per_value = self.bits_per_value;
+        let uncompressed_bits_per_value = self.uncompressed_bits_per_value;
+        async move {
+            let bytes = bytes.await?;
+            Ok(Box::new(BitpackedPageDecoder {
+                bits_per_value,
+                uncompressed_bits_per_value,
+                data: bytes,
+            }) as Box<dyn PhysicalPageDecoder>)
+        }
+        .boxed()
+    }
+}
+
+struct BitpackedPageDecoder {
+    // the number of bits used to represent a compressed value. E.g. if the max value
+    // in the page was 7 (0b111), then this will be 3
+    bits_per_value: u64,
+
+    // number of bits in the uncompressed value. E.g. this will be 32 for u32
+    uncompressed_bits_per_value: u64,
+
+    data: Vec<Bytes>,
+}
+
+impl PhysicalPageDecoder for BitpackedPageDecoder {
+    fn update_capacity(
+        &self,
+        // TODO handle rows to skip
+        _rows_to_skip: u32,
+        num_rows: u32,
+        buffers: &mut [(u64, bool)],
+        // TODO handle all nulls
+        _all_null: &mut bool,
+    ) {
+        // TODO -- not sure if this is correct
+        buffers[0].0 = self.uncompressed_bits_per_value / 8 * num_rows as u64;
+        buffers[0].1 = true;
+    }
+
+    fn decode_into(&self, rows_to_skip: u32, num_rows: u32, dest_buffers: &mut [BytesMut]) {
+        let mut bytes_to_skip = rows_to_skip as u64 * self.bits_per_value / 8;
+        let mut rows_taken = 0;
+        let byte_len = self.uncompressed_bits_per_value / 8;
+        let dst = &mut dest_buffers[0];
+        let mut dst_idx = dst.len(); // index for current byte being written to destination buffer
+
+        // pre-add enough capacity to the buffer to hold all the values we're about to put in it
+        let capacity_to_add = dst.capacity() as i64 - dst.len() as i64 + num_rows as i64;
+        if capacity_to_add > 0 {
+            let bytes_to_add =
+                capacity_to_add as usize * self.uncompressed_bits_per_value as usize / 8;
+            dst.extend((0..bytes_to_add).into_iter().map(|_| 0));
+        }
+
+        // create bitmask for source bits
+        let mut mask = 0u64;
+        for _ in 0..self.bits_per_value {
+            mask = mask << 1 | 1;
+        }
```

Comment:
I think this means you have a limit of 64 bits per value.  This is probably fine but you should add a `debug_assert` somewhere verifying this.